[TG-9294][UFC] Fix context-include/exclude jbmc options

antlechner opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->
The specification I was given by @peterschrammel about how `--context-include/exclude` should work is the following:
- `--context-include` and `--context-exclude` both take a prefix of a Java package, class or method as an argument
- Both can be specified several times on the same command line.
- All methods matching a prefix in `--context-include` and not matching a prefix in `--context-exclude` are "included", all other methods are "excluded".
- Included methods are treated just like regular (non-opaque) methods when `--context-include/exclude` aren't used. Equivalently, `--context-include` can be seen as including everything by default (i.e. if not specified on the command line), and `--context-exclude` can be seen as including nothing by default.
- Excluded methods should have all their meta-information filled in when converting the bytecode, but their body (symbol value) should not be loaded and instead be replaced with a "return nondet null or instance of return type" body, just as we do for stubbed methods. Thus excluded methods are between opaque methods and "regular" methods.

The last specification was not implemented properly. We left out some meta-information (e.g. parameter identifiers) and left the body as nil rather than assigning a stub-like body. This PR fixes both parts of this specification.

- [x] Each commit message has a non-empty body, explaining why the change was made. (Except for documentation commit.)
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix. (Two small closely related fixes.)
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

antlechner committed:
Rename method_in_context to method_context

The new name makes it clearer that this field defines a context, not a
method.

antlechner committed:
Move method context check before symbol value assignment

The previous check was right at the beginning of convert_single_method,
meaning that for an "excluded" method, we would have never entered
java_bytecode_convert_method, which assigns more than just the symbol
value (body of a method), e.g. parameter identifiers.
The only information that we want to omit for excluded methods is their
bodies / symbol values. This is why the new check is just before the
assignment of symbol values, making sure that parameter identifiers and
other meta-information of the method are correctly assigned for excluded
methods.

antlechner committed:
Remove "WARNING: no body for function" regexes

These lines appear in the output for methods whose body is empty/nil.
For excluded methods, we now have "return nondet" bodies as we do for
stubs, so the warnings are no longer printed.

antlechner committed:
Add context-include tests for package and absent prefix

The tests in this commit would have all passed without the previous
fixes already and are just added for completeness.
The tested prefixes are a prefix of a package and a prefix that does not
match anything on the classpath.

antlechner committed:
Add tests for properties of excluded methods

The tests from the previous commit show that the right methods are
excluded.
The tests from this commit show that an excluded method has the right
properties: meta-information such as parameter types and identifiers
should be preserved, while the real method should not be loaded and a
stub-like "return nondet" body should be used instead.
The tests from this commit would not have passed without the fixes from
this PR.

antlechner committed:
Move load_goto_model... declaration to end of file

This is consistent with the location of the definition in the cpp file.

antlechner committed:
Add unit tests for preserving meta-information of excluded methods

These unit tests complement the regression tests in
jbmc/regression/jbmc/context-include-exclude
that were added in a previous commit to check for properties of excluded
methods. Some properties, like information about accessibility and final
and static status, is easier to check in unit-style tests.

antlechner committed:
Split load_goto_model_from_java_class into two functions

The new overload allows specifying the command-line options to run jbmc
with. For example, we will be able to specify that --context-include
should be used.

antlechner committed:
Add documentation for context-include/exclude

romainbrenguier commented on jbmc/src/java_bytecode/java_bytecode_convert_method_class.h:
```diff
@@ -61,9 +61,12 @@ class java_bytecode_convert_methodt:public messaget
+  void operator()(
+    const symbolt &class_symbol,
+    const methodt &method,
+    const optionalt<prefix_filtert> method_context)
```
const &

romainbrenguier commented on jbmc/src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -1003,19 +1003,24 @@ bool jbmc_parse_optionst::can_generate_function_body(const irep_idt &name)
+  INVARIANT(
+    symbol,
+    "Tried to generate a stub body for function that is not in the symbol "
+    "table: " +
```
:pick: the invariant message should rather be formulated to state what should normally be true.

antlechner replied:
The commit that this was in has been removed as it wasn't compatible with symex-driven lazy loading. :car:

romainbrenguier commented on src/goto-programs/lazy_goto_functions_map.h:
```diff
@@ -57,12 +57,10 @@ class lazy_goto_functions_mapt final
+  typedef std::function<bool(
+    const irep_idt &function_name,
+    symbol_table_baset &symbol_table,
+    goto_functiont &function)>
```
:pick: `using` is preferred

antlechner replied:
:car:

romainbrenguier commented on jbmc/regression/jbmc/context-include-exclude/test_include_all.desc:
```diff
@@ -7,6 +7,5 @@ Main.class
 .* line 11 assertion at file Main.java line 11 .*: SUCCESS
 .* line 12 assertion at file Main.java line 12 .*: SUCCESS
 --
-WARNING: no body for function .*
```
:question: is this one removed because it is useless, or does the warning now appear?

antlechner replied:
Because it's (probably) useless. There is a test in a later commit that checks that the body of an excluded method is a stub-style "return nondet" body. Do you think it still makes sense to check for `WARNING: ...` as a negative regex everywhere?

romainbrenguier replied:
If there is no way this warning could be produced for any input then it can be removed, otherwise it's probably safer to keep it in just in case.

antlechner replied:
I'm keeping it as a negative regex here now, and also added it as a negative regex in all existing tests and the new tests for exclude package and exclude absent.

romainbrenguier reviewed (commented):
Reviewed up to `Remove "WARNING: no body for function" regexes`

romainbrenguier commented on jbmc/regression/jbmc/context-include-exclude/test_excluded_deleted_original_body.desc:
```diff
@@ -0,0 +1,12 @@
+--context-exclude org.cprover.other --function ExcludedProperties.runtimeReturnType
+^EXIT=10$
+^SIGNAL=0$
+.* line 30 assertion at file ExcludedProperties.java line 30.*: FAILURE
```
:question: there is no assertion line 30, as far as I can see, I mean 28?

antlechner replied:
As discussed, this was a result of a late clang-format run (30 was correct before clang-format) and should be fixed now.

romainbrenguier commented on jbmc/unit/java-testing-utils/load_java_class.cpp:
```diff
@@ -179,20 +179,37 @@ symbol_tablet load_java_class(
-/// See \ref load_goto_model_from_java_class
-/// With the command line configured to disable lazy loading and string
-/// refinement and the language set to be the default java_bytecode language
 goto_modelt load_goto_model_from_java_class(
```
:pick: should be documented

antlechner replied:
It is documented in the header file.

romainbrenguier commented on jbmc/unit/java_bytecode/java_bytecode_language/context_excluded.cpp:
```diff
@@ -0,0 +1,84 @@
+        const auto public_instance_type =
+          type_try_dynamic_cast<java_method_typet>(public_instance_symbol.type);
+        REQUIRE(public_instance_type);
+        REQUIRE(id2string(public_instance_type->get(ID_access)) == "public");
```
:pick: compare to ID_public, and use `get_access` :gem:

romainbrenguier commented on jbmc/unit/java_bytecode/java_bytecode_language/context_excluded.cpp:
```diff
@@ -0,0 +1,84 @@
+      THEN(
+        "Meta-information of excluded methods is preserved. Symbol values are "
+        "nil as the lazy_goto_modelt for unit tests does not generate "
+        "stub/exclude bodies.")
```
:pick: it would be nice to have several `THEN` sections to structure the test and make it easier to identify what goes wrong when something does.

antlechner replied:
Now grouped into several `THEN` sections where each is about one property of excluded method symbols.

romainbrenguier commented on jbmc/unit/java_bytecode/java_bytecode_language/context_excluded.cpp:
```diff
@@ -0,0 +1,84 @@
+        const auto private_static_type =
+          type_try_dynamic_cast<java_method_typet>(private_static_symbol.type);
+        REQUIRE(private_static_type);
+        REQUIRE(id2string(private_static_type->get(ID_access)) == "private");
```
:gem:

romainbrenguier commented on jbmc/unit/java_bytecode/java_bytecode_language/context_excluded.cpp:
```diff
@@ -0,0 +1,84 @@
+            protected_final_instance_symbol.type);
+        REQUIRE(protected_final_instance_type);
+        REQUIRE(
+          id2string(protected_final_instance_type->get(ID_access)) ==
```
:gem:

romainbrenguier commented on jbmc/unit/java_bytecode/java_bytecode_language/context_excluded.cpp:
```diff
@@ -0,0 +1,84 @@
+            default_final_static_symbol.type);
+        REQUIRE(default_final_static_type);
+        REQUIRE(
+          id2string(default_final_static_type->get(ID_access)) == "default");
```
:gem:

romainbrenguier commented on jbmc/unit/java_bytecode/java_bytecode_language/context_excluded.cpp:
```diff
@@ -0,0 +1,84 @@
+        "stub/exclude bodies.")
+      {
+        const auto &public_instance_symbol = symbol_table.lookup_ref(
+          "java::ClassWithDifferentModifiers.testPublicInstance:()I");
```
:pick: would be better to use `lookup` and `REQUIRE` the result to be non-null :shark:

antlechner replied:
Done. The `REQUIRE`s had to go into the `WHEN` part, is that ok?

romainbrenguier commented on jbmc/unit/java_bytecode/java_bytecode_language/context_excluded.cpp:
```diff
@@ -0,0 +1,84 @@
+        REQUIRE_FALSE(public_instance_type->get_is_final());
+
+        const auto &private_static_symbol = symbol_table.lookup_ref(
+          "java::ClassWithDifferentModifiers.testPrivateStatic:()I");
```
:shark:

romainbrenguier commented on jbmc/unit/java_bytecode/java_bytecode_language/context_excluded.cpp:
```diff
@@ -0,0 +1,84 @@
+        REQUIRE_FALSE(private_static_type->get_is_final());
+
+        const auto &protected_final_instance_symbol = symbol_table.lookup_ref(
+          "java::ClassWithDifferentModifiers.testProtectedFinalInstance:()I");
```
:shark:

romainbrenguier commented on jbmc/unit/java_bytecode/java_bytecode_language/context_excluded.cpp:
```diff
@@ -0,0 +1,84 @@
+        REQUIRE(protected_final_instance_type->get_is_final());
+
+        const auto &default_final_static_symbol = symbol_table.lookup_ref(
+          "java::ClassWithDifferentModifiers.testDefaultFinalStatic:()I");
```
:shark:

antlechner commented:
It turns out my second fix (for assigning stub-like bodies to excluded methods) is too general and incompatible with symex-driven lazy loading. I will add an "in progress" label, and remove it once I've finished fixing that commit.

antlechner committed:
Amend condition for when stub method body should be made

There are three types of method symbols that we expect here:
1) symbols for regular methods, i.e. the bytecode of the method is
available and the method is in the supplied context
2) symbols for opaque methods, i.e. the bytecode of the method is not
available
3) symbols for excluded methods, i.e. the bytecode of the method is
available but the method is not in the supplied context
The previous check was correct for 1) and 2): we do not want to create
stub bodies ("return nondet") for regular methods but we do want to
create them for opaque methods.
The new check makes sure that stub bodies are also created for excluded
methods (3). Previously the value of those symbols was just left as nil,
meaning the (type of the) return value of the method was not constrained
in any way.

antlechner committed:
Remove unused variable assignment

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: cbf532086459a9fcee6266fbb6c364304dad4ac0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/128352945

antlechner committed:
Move convert call above body_provided check

The previous version would have not called convert_single_method on
excluded methods, meaning they would have missed some meta-information,
e.g. parameter identifiers.

antlechner committed:
Protect needed_lazy_methods access in an if-block

If we are not using CI lazy methods here (as could happen after the
change from the previous commit that makes sure convert_single_method is
run for all methods), needed_lazy_methods is an empty optional and we
cannot access it.

antlechner commented:
@romainbrenguier I addressed your comments. I also fixed the issue with symex-driven lazy loading. A previous commit has been replaced with three new commits for the fix:
```
Amend condition for when stub method body should be made
Remove unused variable assignment
Move convert call above body_provided check
```
Could you please review these new commits?

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: ceb297d08d845989b4ee865eb676657cd2733aaa).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/128380227

peterschrammel commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -605,8 +606,12 @@ void java_bytecode_convert_methodt::convert(
   if((!m.is_abstract) && (!m.is_native))
   {
     code_blockt code(convert_parameter_annotations(m, method_type));
```
Why is this line outside the if?

owen-mc-diffblue replied:
Yes, it should be inside. Antonia is away for three days, so I've made a follow-up PR to make this change: https://github.com/diffblue/cbmc/pull/5121

peterschrammel commented on jbmc/src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -690,6 +690,8 @@ int jbmc_parse_optionst::get_goto_program(
   std::unique_ptr<abstract_goto_modelt> &goto_model_ptr,
   const optionst &options)
 {
+  if(options.is_set("context-include") || options.is_set("context-exclude"))
```
It's ugly to repeat that here, but probably can't be fixed without cleaning up the language interface.

romainbrenguier reviewed (approved):
@antlechner 
> Could you please review these new commits?

Still looks fine
